cpp/serve/sampler.cc

```diff
@@ -151,13 +153,18 @@ int SampleTopPFromProb(NDArray prob, int unit_offset, double top_p, double unifo
       static_cast<float*>(__builtin_assume_aligned(prob->data, 4)) + (unit_offset * ndata);
   constexpr double one = 1.0f - 1e-5f;
 
+  if (output_prob_dist) {
+    (*output_prob_dist)[unit_offset] = NDArray::Empty({ndata}, prob->dtype, DLDevice{kDLCPU, 0});
+    (*output_prob_dist)[unit_offset].CopyFromBytes(p_prob, ndata * sizeof(float));
+  }
+
```

Comment:
We can avoid multiple allocation for a same offset.
```suggestion
  if (output_prob_dist) {
    if (!(*output_prob_dist)[unit_offset].defined()) {
      (*output_prob_dist)[unit_offset] = NDArray::Empty({ndata}, prob->dtype, DLDevice{kDLCPU, 0});
    }
    (*output_prob_dist)[unit_offset].CopyFromBytes(p_prob, ndata * sizeof(float));
  }

```